src/main/java/com/chainstep/atlas/example/CustomPolicy.java

```diff
@@ -5,27 +5,142 @@
+     }
+  
+    // Returns true if this policy applies to Verifiable Presentations (VPs)
+    public boolean getApplyToVP() {
+        return true;
+     }
+
+    // The main verification method, checks the VC against custom policy
     @NotNull
     protected VerificationPolicyResult doVerify(@NotNull VerifiableCredential vc) {
-        log.info("CustomPolicy");
-        return VerificationPolicyResult.Companion.success();
+        log.info("CustomPolicy is called");
+
+        try {
+            // Extract the "verifiableCredential" property and convert it to JSON string
+            Object object = vc.getProperties().get("verifiableCredential");
+            ObjectWriter ow = new ObjectMapper().writer().withDefaultPrettyPrinter();
+            String jsonString = ow.writeValueAsString(object);
+            //log.info("returning JSON String: " + jsonString);                      
+
+            ObjectMapper objectMapper = new ObjectMapper();
+            
+            // Parse the JSON string into an array of JsonNodes
+            JsonNode rootNode = objectMapper.readTree(jsonString);
+            
+            // Access the first element in the array (the credential itself)
+            JsonNode credentialNode = rootNode.get(0);
+
+            // Check if the credential is signed by AtlasXAuth
+            if(isSignedByAtlasXAuth(credentialNode)){
+                log.info("isSignedByAtlasXAuth true");   
+            }
+            else{
+                log.info("isSignedByAtlasXAuth false! NO access!");
+                return VerificationPolicyResult.Companion.failure();  // Fail if not signed by AtlasXAuth
+            }
+
+            // Check if the credential has the required legal registration number
+            if(isFromLegalRegistrationNumber(credentialNode, legalRegistrationNumString)){
+                log.info("isFromLegalRegistrationNumber true");   
+            }
+            else{
+                log.info("isFromLegalRegistrationNumber false! NO access!");
+                return VerificationPolicyResult.Companion.failure();  // Fail if the legal registration number does not match
+            }
+
+            // Check if the credential has the required role
+            if(isRole(credentialNode, roleToGrantAccessString)){
+                log.info("isRole true");   
+            }
+            else{
+                log.info("isRole false! NO access!");
+                return VerificationPolicyResult.Companion.failure();  // Fail if the role is not "remoteControl"
+            }
+
+            // If all checks pass, return success
+            return VerificationPolicyResult.Companion.success();
+            
+        } catch (Exception e) {
+            log.info("error during parsing credential subject" + e);
+            return VerificationPolicyResult.Companion.failure(new Exception("parsing Error in the CustomPolicy"));
+        }
+
     }
+
+        private boolean isSignedByAtlasXAuth(JsonNode jsonNode ) {
```

Comment:
We do not have to check isSignedByAtlasXAuth because it is already checked by GxCredentialSignedByAuthorityVerificationPolicy. It is used in file AtlasConfiguration.java :
```java
 @Bean
    public AtlasJwtConfigBuilder atlasJwtConfigBuilder() {
        PresentedBySubjectVerificationPolicy presentedPolicy = new PresentedBySubjectVerificationPolicy();
 -->  GxCredentialSignedByAuthorityVerificationPolicy authorityVerificationPolicy = new GxCredentialSignedByAuthorityVerificationPolicy();
        CustomPolicy customPolicy = new CustomPolicy();
        return configurer -> {
            configurer.configure("/proxy.*").all(List.of(authorityVerificationPolicy, presentedPolicy, customPolicy));
        };
    }
```

Reply:
removed, was not aware of that, but you are absolutely right.